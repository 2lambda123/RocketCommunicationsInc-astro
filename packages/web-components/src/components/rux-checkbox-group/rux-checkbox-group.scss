@use '../../common/functional-components/FormFieldMessage/form-field-message.scss';

:host {
    display: block;
}

.hidden,
:host([hidden]) {
    display: none;
}

::slotted(rux-checkbox) {
    display: block;
    //margin-bottom: 0.625rem; //10px
    margin-bottom: var(--spacing-2);
}

::slotted(rux-checkbox:last-of-type) {
    margin-bottom: 0;
}

.rux-label {
    color: var(--color-text-primary);
    font-family: var(--font-body-1-font-family);
    font-size: var(--font-body-1-font-size);
    font-weight: var(--font-body-1-font-weight);
    letter-spacing: var(--font-body-1-letter-spacing);
    line-height: var(--font-body-1-line-height);
    margin-bottom: var(--spacing-input-label-top);
    &__asterisk {
        margin-left: 4px;
    }
}

.rux-checkbox-group {
    display: flex;
    flex-direction: column;
    padding: 1rem;
    width: fit-content;
    box-shadow: var(--color-border-interactive-muted) 0 0 0 1px inset;
    border-radius: var(--radius-base);
    &--invalid {
        box-shadow: var(--color-text-error) 0 0 0 1px inset;
    }
}
