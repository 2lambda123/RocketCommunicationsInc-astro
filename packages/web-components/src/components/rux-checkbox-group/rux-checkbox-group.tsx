import { Prop, Host, Component, h, State, Element } from '@stencil/core'
import { hasSlot } from '../../utils/utils'

/**
 * @slot (default) - The checkbox elements
 * @slot label - The checkbox group label
 * @slot help-text -  the help text
 * @slot error-text -  the error text
 * @part container - The container div of checkbox elements
 * @part error-text - The error text element
 * @part form-field - The form-field wrapper container
 * @part help-text - The help text element
 * @part label - The input label when `label` prop is set
 * @part required - The asterisk when required is true
 */
@Component({
    tag: 'rux-checkbox-group',
    styleUrl: 'rux-checkbox-group.scss',
    shadow: true,
})
export class RuxCheckboxGroup {
    @Element() el!: HTMLRuxCheckboxGroupElement

    @State() hasLabelSlot = false
    @State() hasHelpSlot = false
    @State() hasErrorSlot = false

    /**
     * The label of the checkbox group. For HTML content, use the `label` slot instead.
     */
    @Prop() label?: string

    /**
     * The help or explanation text
     */
    @Prop({ attribute: 'help-text' }) helpText?: string

    /**
     * The validation error text
     */
    @Prop({ attribute: 'error-text' }) errorText?: string

    /**
     * Presentational only. Renders the Checkbox Group as invalid.
     */
    @Prop() invalid: boolean = false

    /**
     * Marks that a selection from the checkbox group is requried.
     */
    @Prop() required: boolean = false

    connectedCallback() {
        this._handleSlotChange = this._handleSlotChange.bind(this)
    }

    disconnectedCallback() {
        this.el!.shadowRoot!.removeEventListener(
            'slotchange',
            this._handleSlotChange
        )
    }
    componentWillLoad() {
        this._handleSlotChange()
    }

    get hasLabel() {
        return this.label ? true : this.hasLabelSlot
    }

    private _handleSlotChange() {
        this.hasLabelSlot = hasSlot(this.el, 'label')
        this.hasErrorSlot = hasSlot(this.el, 'error-text')
        this.hasHelpSlot = hasSlot(this.el, 'help-text')
    }

    render() {
        const {
            errorText,
            helpText,
            hasHelpSlot,
            hasErrorSlot,
            _handleSlotChange,
        } = this
        return (
            <Host>
                <div class="rux-form-field" part="form-field">
                    <div
                        class={{
                            'rux-label': true,
                            hidden: !this.hasLabel,
                        }}
                        part="label"
                    >
                        <slot
                            onSlotchange={this._handleSlotChange}
                            name="label"
                        >
                            {this.label}
                            {this.required && (
                                <span
                                    part="required"
                                    class="rux-label__asterisk"
                                >
                                    &#42;
                                </span>
                            )}
                        </slot>
                    </div>
                    <div
                        class={{
                            'rux-checkbox-group': true,
                            'rux-checkbox-group--invalid': this.invalid,
                        }}
                        part="container"
                    >
                        <slot></slot>
                    </div>
                </div>
                <div
                    class={{
                        'rux-error-text': !!errorText || hasErrorSlot,
                        hidden: !errorText && !hasErrorSlot,
                    }}
                    part="error-text"
                >
                    <svg
                        fill="none"
                        width="14"
                        height="14"
                        xmlns="http://www.w3.org/2000/svg"
                        viewBox="0 0 14 14"
                    >
                        <path
                            fill-rule="evenodd"
                            clip-rule="evenodd"
                            d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z"
                            fill="currentColor"
                        />
                    </svg>
                    <slot name="error-text" onSlotchange={_handleSlotChange}>
                        {errorText}
                    </slot>
                </div>
                <div
                    class={{
                        'rux-help-text':
                            (!!helpText || hasHelpSlot) &&
                            (!errorText || !hasErrorSlot),
                        hidden:
                            (!helpText && !hasHelpSlot) ||
                            !!errorText ||
                            hasErrorSlot,
                    }}
                    part="help-text"
                >
                    <slot name="help-text" onSlotchange={_handleSlotChange}>
                        {helpText}
                    </slot>
                </div>
            </Host>
        )
    }
}
