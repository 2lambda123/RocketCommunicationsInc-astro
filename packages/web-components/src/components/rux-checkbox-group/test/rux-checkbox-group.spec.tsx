import { newSpecPage } from '@stencil/core/testing'
import { RuxCheckboxGroup } from '../rux-checkbox-group'

describe('rux-checkbox-group', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxCheckboxGroup],
            html: `<rux-checkbox-group></rux-checkbox-group>`,
        })
        expect(page.root).toMatchSnapshot()
    })
    it('renders label prop', async () => {
        const page = await newSpecPage({
            components: [RuxCheckboxGroup],
            html: `<rux-checkbox-group label="hello"></rux-checkbox-group>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('renders label slot', async () => {
        const page = await newSpecPage({
            components: [RuxCheckboxGroup],
            html: `<rux-checkbox-group><div slot="label">hello</div></rux-checkbox-group>`,
        })
        expect(page.root).toMatchSnapshot()
    })
})
