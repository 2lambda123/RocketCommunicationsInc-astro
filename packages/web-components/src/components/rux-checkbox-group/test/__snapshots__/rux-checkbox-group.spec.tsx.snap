// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-checkbox-group renders 1`] = `
<rux-checkbox-group>
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <div class="hidden rux-label" part="label">
        <slot name="label"></slot>
      </div>
      <div class="rux-checkbox-group" part="container">
        <slot></slot>
      </div>
    </div>
    <div class="hidden" part="error-text">
      <svg fill="none" height="14" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
        <path clip-rule="evenodd" d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z" fill="currentColor" fill-rule="evenodd"></path>
      </svg>
      <slot name="error-text"></slot>
    </div>
    <div class="hidden" part="help-text">
      <slot name="help-text"></slot>
    </div>
  </mock:shadow-root>
</rux-checkbox-group>
`;

exports[`rux-checkbox-group renders label prop 1`] = `
<rux-checkbox-group label="hello">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <div class="rux-label" part="label">
        <slot name="label">
          hello
        </slot>
      </div>
      <div class="rux-checkbox-group" part="container">
        <slot></slot>
      </div>
    </div>
    <div class="hidden" part="error-text">
      <svg fill="none" height="14" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
        <path clip-rule="evenodd" d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z" fill="currentColor" fill-rule="evenodd"></path>
      </svg>
      <slot name="error-text"></slot>
    </div>
    <div class="hidden" part="help-text">
      <slot name="help-text"></slot>
    </div>
  </mock:shadow-root>
</rux-checkbox-group>
`;

exports[`rux-checkbox-group renders label slot 1`] = `
<rux-checkbox-group>
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <div class="rux-label" part="label">
        <slot name="label"></slot>
      </div>
      <div class="rux-checkbox-group" part="container">
        <slot></slot>
      </div>
    </div>
    <div class="hidden" part="error-text">
      <svg fill="none" height="14" viewBox="0 0 14 14" width="14" xmlns="http://www.w3.org/2000/svg">
        <path clip-rule="evenodd" d="M11.393 12.25c.898 0 1.458-.974 1.009-1.75L8.009 2.91a1.166 1.166 0 0 0-2.018 0L1.598 10.5c-.449.776.111 1.75 1.01 1.75h8.784ZM7 8.167a.585.585 0 0 1-.583-.584V6.417c0-.321.262-.584.583-.584.32 0 .583.263.583.584v1.166c0 .321-.262.584-.583.584Zm-.583 1.166V10.5h1.166V9.333H6.417Z" fill="currentColor" fill-rule="evenodd"></path>
      </svg>
      <slot name="error-text"></slot>
    </div>
    <div class="hidden" part="help-text">
      <slot name="help-text"></slot>
    </div>
  </mock:shadow-root>
  <div slot="label">
    hello
  </div>
</rux-checkbox-group>
`;
